File: packages/platform_ios_tts/ios/Classes/engines/SupertonicTtsService.swift
import Foundation

/// Supertonic TTS engine service.
/// Uses ONNX Runtime for 4-stage pipeline synthesis (matches official Supertonic SDK).
class SupertonicTtsService: TtsServiceProtocol {
    let engineType: NativeEngineType = .supertonic
    
    private var onnxInference = SupertonicOnnxInference()
    private var loadedVoices: [String: VoiceInfo] = [:]
    private let lock = NSLock()
    private let synthesisCounter = SynthesisCounter()
    
    var isReady: Bool {
        lock.lock()
        defer { lock.unlock() }
        return onnxInference.isModelLoaded
    }
    
    func loadCore(corePath: String, configPath: String?) async throws {
        // Thread logging: loadCore entry
        NSLog("[THREAD] SupertonicTtsService.loadCore ENTRY: isMainThread=%@, thread=%@", 
              Thread.isMainThread ? "YES" : "NO", 
              Thread.current.description)
        NSLog("[SupertonicTtsService] loadCore called with corePath: %@", corePath)
        
        // Check if already loaded (quick check without heavy work)
        lock.lock()
        let alreadyLoaded = onnxInference.isModelLoaded
        lock.unlock()
        
        if alreadyLoaded {
            NSLog("[SupertonicTtsService] ONNX models already loaded")
            return
        }
        
        // Load ONNX models on a background thread
        let inference = self.onnxInference
        let path = corePath
        try await Task.detached(priority: .userInitiated) {
            // Thread logging: Inside Task.detached for model loading
            NSLog("[THREAD] SupertonicTtsService.loadCore Task.detached: isMainThread=%@, thread=%@", 
                  Thread.isMainThread ? "YES" : "NO", 
                  Thread.current.description)
            NSLog("[SupertonicTtsService] Loading ONNX models from background thread: %@", path)
            try inference.loadFromDirectory(path)
            NSLog("[SupertonicTtsService] ONNX models loaded successfully")
        }.value
    }
    
    func loadVoice(voiceId: String, modelPath: String, speakerId: Int?, configPath: String?) async throws {
        lock.lock()
        defer { lock.unlock() }
        
        // For Supertonic, all voices use the same model with different speakerIds
        // Just register the voice - model is loaded during loadCore
        loadedVoices[voiceId] = VoiceInfo(
            voiceId: voiceId,
            modelPath: modelPath,
            speakerId: speakerId,
            lastUsed: Date()
        )
        NSLog("[SupertonicTtsService] Voice registered: %@ with speakerId: %d", voiceId, speakerId ?? 0)
    }
    
    func synthesize(text: String, voiceId: String, outputPath: String, speakerId: Int?, speed: Double) async throws -> SynthesizeResult {
        // Track active synthesis to prevent unload during operation
        synthesisCounter.increment()
        defer { synthesisCounter.decrement() }
        
        lock.lock()
        let voice = loadedVoices[voiceId]
        lock.unlock()
        
        guard voice != nil else {
            throw TtsError.voiceNotLoaded(voiceId)
        }
        
        guard !text.isEmpty else {
            throw TtsError.invalidInput("Text is empty")
        }
        
        // Thread logging: Entry point (should be background from PlatformIosTtsPlugin)
        NSLog("[THREAD] SupertonicTtsService.synthesize ENTRY: isMainThread=%@, thread=%@", 
              Thread.isMainThread ? "YES" : "NO", 
              Thread.current.description)
        
        NSLog("[SupertonicTtsService] Synthesizing with ONNX: voiceId=%@, speakerId=%d, text length=%d", voiceId, speakerId ?? 0, text.count)
        NSLog("[SupertonicTtsService] isModelLoaded=%@", onnxInference.isModelLoaded ? "YES" : "NO")
        
        // Run synthesis on background thread with lower priority to avoid UI contention
        let inference = self.onnxInference
        let (samples, sampleRate, elapsed) = try await Task.detached(priority: .utility) {
            // Thread logging: Inside utility priority Task (should be background)
            NSLog("[THREAD] SupertonicTtsService ONNX inference: isMainThread=%@, thread=%@", 
                  Thread.isMainThread ? "YES" : "NO", 
                  Thread.current.description)
            
            let startTime = CFAbsoluteTimeGetCurrent()
            let (samples, sampleRate) = try inference.synthesize(
                text: text,
                voiceName: voiceId,
                speakerId: speakerId ?? 0,
                speed: Float(speed)
            )
            let elapsed = CFAbsoluteTimeGetCurrent() - startTime
            return (samples, sampleRate, elapsed)
        }.value
        
        // Thread logging: After Task.detached returns
        NSLog("[THREAD] SupertonicTtsService after ONNX: isMainThread=%@, thread=%@", 
              Thread.isMainThread ? "YES" : "NO", 
              Thread.current.description)
        
        NSLog("[SupertonicTtsService] Synthesis complete: %d samples at %dHz in %.2fs", samples.count, sampleRate, elapsed)
        
        // Write WAV file
        try AudioConverter.writeWav(samples: samples, sampleRate: sampleRate, to: outputPath)
        
        let durationMs = AudioConverter.durationMs(sampleCount: samples.count, sampleRate: sampleRate)
        
        return SynthesizeResult(
            success: true,
            durationMs: durationMs,
            sampleRate: Int64(sampleRate),
            errorCode: .none,
            errorMessage: nil
        )
    }
    
    func cancelSynthesis(requestId: String) {
        NSLog("[SupertonicTtsService] Cancel requested for: %@", requestId)
    }
    
    func unloadVoice(voiceId: String) {
        // Wait for active synthesis operations to complete
        _ = synthesisCounter.waitUntilIdle(timeoutMs: 5000)
        
        lock.lock()
        defer { lock.unlock() }
        loadedVoices.removeValue(forKey: voiceId)
        onnxInference.unload()
        NSLog("[SupertonicTtsService] Voice unloaded: %@", voiceId)
    }
    
    func unloadAll() {
        // Wait for active synthesis operations to complete
        _ = synthesisCounter.waitUntilIdle(timeoutMs: 5000)
        
        lock.lock()
        defer { lock.unlock() }
        loadedVoices.removeAll()
        onnxInference.unload()
        NSLog("[SupertonicTtsService] All resources unloaded")
    }
}

File: packages/platform_android_tts/android/src/main/kotlin/com/example/platform_android_tts/services/SupertonicTtsService.kt
package com.example.platform_android_tts.services

import android.app.Service
import android.content.Intent
import android.os.IBinder
import com.example.platform_android_tts.onnx.SupertonicNative
import kotlinx.coroutines.*
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Semaphore

/**
 * Supertonic TTS Service running in its own process (:supertonic).
 * 
 * Provides synthesis using Supertonic ONNX TTS models via JNI.
 * Uses speaker embeddings for voice cloning/variation.
 */
class SupertonicTtsService : Service() {
    
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    
    // Synthesis counter to prevent unload during active synthesis
    private val synthesisCounter = SynthesisCounter()
    
    // Model state
    private var modelPath: String? = null
    @Volatile private var isInitialized = false
    private val loadedSpeakers = ConcurrentHashMap<String, SupertonicSpeaker>()
    
    // Active synthesis jobs for cancellation (thread-safe)
    private val activeJobs = ConcurrentHashMap<String, Job>()
    
    // Limit concurrent synthesis to prevent resource exhaustion
    private val synthesisPermits = Semaphore(4)
    
    override fun onBind(intent: Intent?): IBinder? = null
    
    override fun onDestroy() {
        scope.cancel()
        unloadAllModels()
        // Dispose native resources
        if (isInitialized) {
            SupertonicNative.dispose()
        }
        super.onDestroy()
    }
    
    /**
     * Initialize the Supertonic engine with the core model path.
     * 
     * Expected structure:
     * corePath/
     *   onnx/
     *     text_encoder.onnx
     *     duration_predictor.onnx
     *     vector_estimator.onnx
     *     vocoder.onnx
     *     unicode_indexer.json
     *   voice_styles/
     *     M1.json, M2.json, ..., F1.json, F2.json, ...
     */
    suspend fun initEngine(corePath: String): Result<Unit> = runCatching {
        if (isInitialized && modelPath == corePath) {
            return Result.success(Unit)
        }
        
        // Check for native library availability
        if (!SupertonicNative.isNativeAvailable()) {
            throw IllegalStateException("Supertonic native library not available")
        }
        
        // Verify model files exist
        val textEncoder = File(corePath, "onnx/text_encoder.onnx")
        if (!textEncoder.exists()) {
            throw IllegalStateException("Model file not found: ${textEncoder.path}")
        }
        
        // Initialize native engine
        val success = SupertonicNative.initialize(corePath)
        if (!success) {
            throw IllegalStateException("Failed to initialize Supertonic native engine at: $corePath")
        }
        
        modelPath = corePath
        isInitialized = true
        android.util.Log.i("SupertonicTtsService", "Engine initialized at: $corePath")
    }
    
    /**
     * Load a speaker with optional embedding.
     */
    suspend fun loadVoice(
        voiceId: String, 
        speakerId: Int,
        embeddingPath: String? = null
    ): Result<Unit> = runCatching {
        if (!isInitialized) {
            throw IllegalStateException("Engine not initialized")
        }
        
        // TODO: Load speaker embedding if provided
        loadedSpeakers[voiceId] = SupertonicSpeaker(
            voiceId = voiceId,
            speakerId = speakerId,
            embeddingPath = embeddingPath,
            lastUsed = System.currentTimeMillis()
        )
    }
    
    /**
     * Synthesize text to a WAV file using Supertonic native engine.
     */
    suspend fun synthesize(
        voiceId: String,
        text: String,
        outputPath: String,
        requestId: String,
        speakerId: Int = 0,
        speed: Float = 1.0f
    ): SynthesisResult {
        if (!isInitialized || !SupertonicNative.isReady()) {
            return SynthesisResult(
                success = false,
                errorCode = ErrorCode.MODEL_MISSING,
                errorMessage = "Engine not initialized"
            )
        }
        
        if (text.isBlank()) {
            return SynthesisResult(
                success = false,
                errorCode = ErrorCode.INVALID_INPUT,
                errorMessage = "Empty text"
            )
        }
        
        // Track active synthesis to prevent unload during operation
        synthesisCounter.increment()
        
        // Limit concurrent synthesis
        if (!synthesisPermits.tryAcquire()) {
            synthesisCounter.decrement()
            return SynthesisResult(
                success = false,
                errorCode = ErrorCode.BUSY,
                errorMessage = "Too many concurrent synthesis requests"
            )
        }
        
        // Get or create speaker
        val speaker = loadedSpeakers.getOrPut(voiceId) {
            SupertonicSpeaker(
                voiceId = voiceId,
                speakerId = speakerId,
                lastUsed = System.currentTimeMillis()
            )
        }
        
        var audioSamples: FloatArray? = null
        var synthError: Exception? = null
        
        val job = scope.launch {
            try {
                // Run native ONNX inference
                audioSamples = SupertonicNative.synthesize(text, speaker.speakerId, speed)
                
                if (audioSamples == null) {
                    synthError = IllegalStateException("Native synthesis returned null")
                    return@launch
                }
                
                ensureActive()
                
                // Convert float samples to 16-bit PCM and write WAV
                val sampleRate = SupertonicNative.getSampleRate()
                val pcmData = floatToPcm16(audioSamples!!)
                
                val tmpFile = File("$outputPath.tmp")
                val parentDir = tmpFile.parentFile
                if (parentDir != null && !parentDir.exists() && !parentDir.mkdirs()) {
                    throw IOException("Failed to create output directory: $parentDir")
                }
                writeWavFile(tmpFile, pcmData, sampleRate)
                
                ensureActive()
                val finalFile = File(outputPath)
                if (!tmpFile.renameTo(finalFile)) {
                    // Fallback: copy and delete
                    tmpFile.copyTo(finalFile, overwrite = true)
                    tmpFile.delete()
                }
                
            } catch (e: CancellationException) {
                File("$outputPath.tmp").delete()
                throw e
            } catch (e: Exception) {
                synthError = e
                android.util.Log.e("SupertonicTtsService", "Synthesis error", e)
            }
        }
        
        activeJobs[requestId] = job
        
        return try {
            job.join()
            
            if (job.isCancelled) {
                SynthesisResult(
                    success = false,
                    errorCode = ErrorCode.CANCELLED,
                    errorMessage = "Synthesis cancelled"
                )
            } else if (synthError != null) {
                SynthesisResult(
                    success = false,
                    errorCode = mapExceptionToErrorCode(synthError!!),
                    errorMessage = synthError!!.message
                )
            } else if (audioSamples != null) {
                speaker.lastUsed = System.currentTimeMillis()
                val sampleRate = SupertonicNative.getSampleRate()
                val durationMs = (audioSamples!!.size * 1000 / sampleRate)
                
                SynthesisResult(
                    success = true,
                    durationMs = durationMs,
                    sampleRate = sampleRate
                )
            } else {
                SynthesisResult(
                    success = false,
                    errorCode = ErrorCode.INFERENCE_FAILED,
                    errorMessage = "No audio output generated"
                )
            }
        } catch (e: Exception) {
            SynthesisResult(
                success = false,
                errorCode = mapExceptionToErrorCode(e),
                errorMessage = e.message
            )
        } finally {
            activeJobs.remove(requestId)
            synthesisCounter.decrement()
            synthesisPermits.release()
        }
    }
    
    /**
     * Cancel an in-flight synthesis.
     */
    fun cancelSynthesis(requestId: String) {
        // Remove first, then cancel (prevents race with completion)
        val job = activeJobs.remove(requestId) ?: return
        job.cancel()
    }
    
    /**
     * Unload a specific voice.
     * Waits for any active synthesis to complete first.
     */
    fun unloadVoice(voiceId: String) {
        // Wait for any active synthesis to complete (max 5 seconds)
        if (!synthesisCounter.waitUntilIdle(timeoutMs = 5000)) {
            android.util.Log.w("SupertonicTtsService", "Timeout waiting for synthesis to complete before unload")
        }
        loadedSpeakers.remove(voiceId)
    }
    
    /**
     * Unload all models and reset state.
     * Waits for any active synthesis to complete first.
     */
    fun unloadAllModels() {
        // Wait for any active synthesis to complete (max 5 seconds)
        if (!synthesisCounter.waitUntilIdle(timeoutMs = 5000)) {
            android.util.Log.w("SupertonicTtsService", "Timeout waiting for synthesis to complete before unloadAll")
        }
        
        loadedSpeakers.clear()
        
        if (isInitialized) {
            SupertonicNative.dispose()
        }
        
        isInitialized = false
        modelPath = null
        
        activeJobs.values.forEach { it.cancel() }
        activeJobs.clear()
    }
    
    /**
     * Unload least recently used speaker.
     */
    fun unloadLeastUsedVoice(): String? {
        val lru = loadedSpeakers.minByOrNull { it.value.lastUsed }
        return lru?.let {
            loadedSpeakers.remove(it.key)
            it.key
        }
    }
    
    /**
     * Get memory info.
     */
    fun getMemoryInfo(): ServiceMemoryInfo {
        val runtime = Runtime.getRuntime()
        val availableMB = ((runtime.maxMemory() - runtime.totalMemory()) / (1024 * 1024)).toInt()
        val totalMB = (runtime.maxMemory() / (1024 * 1024)).toInt()
        
        return ServiceMemoryInfo(
            availableMB = availableMB,
            totalMB = totalMB,
            loadedModelCount = loadedSpeakers.size
        )
    }
    
    fun isReady(): Boolean = isInitialized && SupertonicNative.isReady()
    fun isVoiceLoaded(voiceId: String): Boolean = loadedSpeakers.containsKey(voiceId)
    
    // Private helpers
    
    /**
     * Convert float audio samples [-1.0, 1.0] to 16-bit PCM.
     */
    private fun floatToPcm16(samples: FloatArray): ShortArray {
        return ShortArray(samples.size) { i ->
            val sample = samples[i].coerceIn(-1.0f, 1.0f)
            (sample * 32767).toInt().toShort()
        }
    }
    
    private fun estimateDuration(text: String): Float {
        val words = text.length / 5.0f
        return words / 2.5f
    }
    
    private fun estimateDurationMs(text: String): Int {
        return (estimateDuration(text) * 1000).toInt()
    }
    
    private fun writeWavFile(file: File, samples: ShortArray, sampleRate: Int) {
        val numChannels = 1
        val bitsPerSample = 16
        val byteRate = sampleRate * numChannels * bitsPerSample / 8
        val blockAlign = numChannels * bitsPerSample / 8
        val dataSize = samples.size * 2
        val fileSize = 36 + dataSize
        
        RandomAccessFile(file, "rw").use { raf ->
            raf.writeBytes("RIFF")
            raf.writeIntLE(fileSize)
            raf.writeBytes("WAVE")
            raf.writeBytes("fmt ")
            raf.writeIntLE(16)
            raf.writeShortLE(1)
            raf.writeShortLE(numChannels)
            raf.writeIntLE(sampleRate)
            raf.writeIntLE(byteRate)
            raf.writeShortLE(blockAlign)
            raf.writeShortLE(bitsPerSample)
            raf.writeBytes("data")
            raf.writeIntLE(dataSize)
            
            val buffer = ByteBuffer.allocate(samples.size * 2).order(ByteOrder.LITTLE_ENDIAN)
            samples.forEach { buffer.putShort(it) }
            raf.write(buffer.array())
        }
    }
    
    private fun RandomAccessFile.writeIntLE(value: Int) {
        val buffer = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN)
        buffer.putInt(value)
        write(buffer.array())
    }
    
    private fun RandomAccessFile.writeShortLE(value: Int) {
        val buffer = ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN)
        buffer.putShort(value.toShort())
        write(buffer.array())
    }
    
    private fun mapExceptionToErrorCode(e: Exception): ErrorCode {
        return when {
            e is OutOfMemoryError -> ErrorCode.OUT_OF_MEMORY
            e is CancellationException -> ErrorCode.CANCELLED
            e.message?.contains("model", ignoreCase = true) == true -> ErrorCode.MODEL_MISSING
            else -> ErrorCode.INFERENCE_FAILED
        }
    }
}

/**
 * Supertonic speaker state.
 */
data class SupertonicSpeaker(
    val voiceId: String,
    val speakerId: Int,
    val embeddingPath: String? = null,
    var lastUsed: Long
)
